tests/functional/pyocf/types/volume_replicated.py

```diff
@@ -0,0 +1,100 @@
+#
+# Copyright(c) 2022 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+from threading import Lock
+from .volume import Volume, VOLUME_POISON
+from .io import Io, IoDir
+from ctypes import cast, c_void_p, CFUNCTYPE, c_int, POINTER, memmove, sizeof, pointer
+
+
+class ReplicatedVolume(Volume):
+    def __init__(self, primary: Volume, secondary: Volume, uuid=None):
+        super().__init__(uuid)
+        self.primary = primary
+        self.secondary = secondary
+
+        if secondary.get_max_io_size() < primary.get_max_io_size():
+            raise Exception("secondary volume max io size too small")
+        if secondary.get_length() < primary.get_length():
+            raise Exception("secondary volume size too small")
+
+    def open(self):
+        ret = self.primary.open()
+        if ret:
+            return ret
+        ret = self.secondary.open()
+        if ret:
+            self.primary.close()
+        return ret
+
+    def close(self):
+        self.primary.close()
+        self.secondary.close()
+
+    def get_length(self):
+        return self.primary.get_length()
+
+    def get_max_io_size(self):
+        return self.primary.get_max_io_size()
+
+    def _prepare_io(self, io):
+        original_cb = Io.END()
+        pointer(original_cb)[0] = io.contents._end
+        lock = Lock()
+        error = 0
+        io_remaining = 2
+
+        @CFUNCTYPE(None, c_void_p, c_int)
+        def cb(io, err):
+            nonlocal io_remaining
+            nonlocal error
+            nonlocal original_cb
+            nonlocal lock
+            io = cast(io, POINTER(Io))
+
+            lock.acquire()
```

Comment:
I think it would be more readable as context manager:
```python
 with lock:
            if err:
                error = err
            io_remaining -= 1
            finished = True if io_remaining == 0 else False
```

Reply:
done